File: frontend/src/types/regime_detection.ts
import type { JobStatus } from './common'

export type { JobStatus }
export type ModelType = 'threshold' | 'vol_quantile' | 'hmm'

export interface ModelConfig {
  model_id: string
  model_type: ModelType
  params: Record<string, unknown>
}

export interface RegimeDetectionRequest {
  assets: string[]
  freq: string
  start_date?: string
  end_date?: string
  models: ModelConfig[]
  prob_window: number
}

export interface RegimeModelResult {
  model_id: string
  model_type: string
  unique_regimes: string[]
  labels: [string, string][]
  probabilities: Record<string, [string, number][]>
}

export interface RegimeDetectionResult {
  assets: string[]
  freq: string
  date_range: [string, string]
  prices: Record<string, [string, number][]>
  volumes: Record<string, [string, number][]>
  models: RegimeModelResult[]
}

export interface RegimeDetectionJobResponse {
  job_id: string
  status: JobStatus
  result?: RegimeDetectionResult
  error?: string
}

File: backend/back_api/src/back_api/schemas/regime_detection.py
from __future__ import annotations

from typing import Literal

from pydantic import BaseModel


class ModelConfig(BaseModel):
    model_id: str
    model_type: Literal["threshold", "vol_quantile", "hmm"]
    params: dict = {}


class RegimeDetectionRequest(BaseModel):
    assets: list[str] = ["BTC", "ETH"]
    freq: str = "1D"
    start_date: str | None = None
    end_date: str | None = None
    models: list[ModelConfig] = []
    prob_window: int = 20


class RegimeModelResult(BaseModel):
    model_id: str
    model_type: str
    unique_regimes: list[str]
    labels: list[tuple[str, str]]
    probabilities: dict[str, list[tuple[str, float]]]


class RegimeDetectionResult(BaseModel):
    assets: list[str]
    freq: str
    date_range: tuple[str, str]
    prices: dict[str, list[tuple[str, float]]]
    volumes: dict[str, list[tuple[str, float]]]
    models: list[RegimeModelResult]


class RegimeDetectionJobResponse(BaseModel):
    job_id: str
    status: str
    result: dict | None = None
    error: str | None = None
